source/slang/hlsl.meta.slang

```diff
@@ -8802,18 +9069,100 @@ void __metalInterlocked_compare_exchange(AtomicType dest, T compare_value, T val
     original_value = compare_value;
 }
 
+__generic<T>
 __glsl_version(430)
 [ForceInline]
+[require(cuda)]
+void __cudaInterlocked_compare_exchange(__ref T dest, __ref T compare_value, T value)
+{
+    __intrinsic_asm "atomicCAS($0, $1, $2)";
+}
+
+__generic<T>
+[ForceInline]
+[require(cuda)]
+void __cudaInterlocked_compare_exchange(__ref T dest, T compare_value, T value, out T original_value)
+{
+    __intrinsic_asm "*$3 = atomicCAS($0, $1, $2)";
+}
+
+__generic<T>
+[ForceInline]
+[require(glsl)]
+void __glslInterlocked_compare_exchange(__ref T dest, __ref T compare_value, T value)
+{
+    __intrinsic_asm "$atomicCompSwap($A, $1, $2)";
+}
+
+__generic<T>
+[ForceInline]
+[require(glsl)]
+void __glslInterlocked_compare_exchange(__ref T dest, T compare_value, T value, out T original_value)
+{
+    __intrinsic_asm "($3 = $atomicCompSwap($A, $1, $2))";
+}
+
+__generic<T>
+[ForceInline]
+[require(hlsl)]
+void __hlslInterlocked_compare_exchange(__ref T dest, __ref T compare_value, T value)
+{
+    __intrinsic_asm "InterlockedCompareExchange";
+}
+
+__generic<T>
+[ForceInline]
+[require(hlsl)]
+void __hlslInterlocked_compare_exchange(__ref T dest, T compare_value, T value, out T original_value)
+{
+    __intrinsic_asm "InterlockedCompareExchange";
+}
+
+__generic<T>
+[ForceInline]
+[require(hlsl)]
+void __hlslInterlocked_compare_exchange_float_bitwise(__ref T dest, T compare_value, T value)
+{
+    __intrinsic_asm "InterlockedCompareExchangeFloatBitwise";
+}
+
+__generic<T>
+[ForceInline]
+[require(hlsl)]
+void __hlslInterlocked_compare_exchange_float_bitwise(__ref T dest, T compare_value, T value, out T original_value)
+{
+    __intrinsic_asm "InterlockedCompareExchangeFloatBitwise";
+}
+
+[ForceInline]
+__glsl_version(430)
```

Comment:
We should turn this into a meta programming loop for all atomic operations to remove code duplication, just like `__hlslInterlocked_add` etc.

We need to add a comment before the meta loop to list the full intrinsic function names the loop expands into, so that we can search for a intrinsic function name in the source file and locate the meta loop.